### PR TITLE
Make --incremental compile distinct .c files for every module

### LINKED ISSUE #19377: What would it take to make `--incremental` even more fine-grained?
I'd always imagined that the `--incremental` flag compiled each generated C file separately, where we generate approximately one C file per Chapel module.  However, it turns out that only the user modules are teased apart and that most of the other (standard and internal) modules are bundled into `_main.c`, making it the bottleneck for many compilations.  This seems particularly noticeable if using `make -j` on the output as explored here:  https://github.com/chapel-lang/chapel/pull/19369

This issue asks whether we could make the `--incremental` flag even more fine-grained by having further decomposing `_main.c` into its component modules.

### PR DESCRIPTION
**Pull request by bradcray**
The --incremental compiler flag generates a distinct C file per *user*
module, but standard and internal modules still get rolled
(conceptually, via #includes) into a single _main.c file.  The result
is that the compilation of the _main.c file tends to be a bottleneck
relative to the other ones.

This PR changes this so that every module is compiled as its own
distinct C file.

While here, I also moved the user error for compiling with
target-compiler=llvm and --incremental earlier in the compiler
so that time isn't wasted before hitting it.

Resolves #19377

### CODE REVIEW

**Comment by bradcray** [comment]
@e-kayrakli : For your interest.

**Review by e-kayrakli — APPROVED** [review]
It might be good to get a green light from someone who worked on `--incremental` for this. It is very plausible that `MOD_USER` limitations were added because of extra caution when first merging this, and then it stayed there forever. But there might be something more involved than that?

**Comment by bradcray** [comment]
Right, I think @lydia-duncan is probably our main person there, though I can't remember who else might have helped with overseeing the project.  I was thinking we should see whether it causes any notable performance degradation for CHAMPS (vs. building with `--incremental` on `main`), as well as how much or little it speeds up compilation in the `make -j` mode.  If performance impact is minimal and compilation time is improved, it seems like a no-brainer.  If performance impact is worse, then maybe we have multiple "incremental" modes which say how incremental you want to be.  And I can't imagine that compilation time won't improve, at least in a `make -j` setting.  If it was significantly hurt in a serial `make` setting (and it may be given the size of chpl__header.h and the number of times it needs to be compiled), that could be another reason to have multiple incremental levels / modes.

**Comment by lydia-duncan** [comment]
I'm the only one still on the team that was involved in the project.

If my memory serves, the intention behind bundled all the non-user modules together was that hopefully they wouldn't be the ones that would be changing, so having them set in one place separated from the modules that are expected to change would be a benefit in the long run.

My main worry about adding multiple incremental modes is that I see this functionality as being greatly improved and subsumed by the new compiler, so any improvements to `--incremental` would be a passing benefit.

**Comment by bradcray** [comment]
> so any improvements to --incremental would be a passing benefit.

That's a very patient / long-view perspective given that users are suffering from (and complaining about) compile times today and a compiler that supports separate compilation / incremental compilation is still arguably years away.  Particularly when the fix literally took me 10 minutes to implement.  :D

**Comment by e-kayrakli** [comment]
Pasting the comment from #19369 because it is also relevant here.

Here's some more CHAMPS context to make this more self contained. We used `incremental` in CHAMPS because they were running out of memory while compiling their application. It turned out that backend was consuming a lot of memory because we were compiling everything as a whole. `incremental` (with a fix I filed in the past) helped solve the memory issue, but it makes the compilation slower. Potentially due to the backend compiler being invoked multiple times, starting IO multiple times etc etc.

With that, here is Chapel compilation time and memory consumption with this PR and its sibling:

| Version (flow executable)                | Front Time  | Front Mem | Back    | Back Mem
|------------------------------------------|-------------|-----------|---------|---------
| main (1)                                 | 2:20.28     | 3.1G      | 3:34.21 | 6.9G
| main + `--incremental` (2)               | 2:20.28     | 3.1G      | 4:51.77 | 1.6G
| main + `--incremental` + #19369          | 2:22.49     | 3.1G      | 2:04.23 | 1.6G
| main + `--incremental` + #19369 + #19407 | 2:36.07     | 3.1G      | 0:52.51 | 0.6G

(1) How CHAMPS is compiled for production today.
(2) How CHAMPS is compiled for development today.

Amazing.

Note that #19407 also reduces the peak memory because it separates the compilation of `main.c`. However, it increases the frontend compilation time likely because it creates more files.

**Comment by e-kayrakli** [comment]
@lydia-duncan is it really passing, though? I thought the only first phases of compilation will be "incremental" with dyno effort IIRC correctly. At least as the first step. Even if that's the case, looking at the benefit, I think we should merge this PR yesterday :)

Nonetheless, maybe for the next release, we can consider packaging this compilation in a different shape because it is not really making incremental work the way it was initially intended. Offline, Brad mentioned maybe making this a `-j` flag in the Chapel compiler. I think we need something like that. But left to my own devices, I'd pick a more verbose name for the flag `--parallel-compile` or something.

**Comment by bradcray** [comment]
@e-kayrakli : A couple of questions:

1) In your table, the final two rows are with `make -j`, is that right?  Did you do any runs to see how long a non-`make -j` compilation will take under this PR, since it creates more `.c` files?

2) Did you do any checks to see how performance compared between rows 3 and 4?

**Comment by bradcray** [comment]
> Offline, Brad mentioned maybe making this a -j flag in the Chapel compiler. I think we need something like that. But left to my own devices, I'd pick a more verbose name for the flag --parallel-compile or something.

This is starting to get into what I was going to open as a separate issue, but I think we should potentially do this for this release, as I think it's very low-hanging fruit and very useful.  It won't actually help the CHAMPS team since they'll probably still want to stop the compilation after the first step, but it would simplify things for others who don't want to use `--savec` and/or `incremental`.

I was imagining a (developer) flag that had both the short-form `-j` and a longer, more descriptive form as well.  Specifically, I'm imagining we'd:

* have throwing this flag enable `--incremental`
* have it complain if used with `--llvm`
* change the `make` call into a `make -j` call

One challenge (which is unfortunate, but not a show-stopper) is that I don't think we have a way to have a flag that supports either an integer or nothing, as `-j` does on `make`.  So, I'd probably have it require an integer value where maybe we'd have non-positive values result in `make -j` and positive values would indicate the `-j` value.

**Comment by e-kayrakli** [comment]
> In your table, the final two rows are with make -j, is that right? Did you do any runs to see how long a non-make -j compilation will take under this PR, since it creates more .c files?

Yes. In fact, all of them are with `-j` but it is meaningless for the first two.

> Did you do any checks to see how performance compared between rows 3 and 4?

I didn't, but can. If I was betting, I would say that the difference wouldn't be a lot, at least not detrimental for a development build. This is not well-informed at all, but I expect the backend optimizations that are more local to impact performance more than whole-program optimization. Besides row 3 is already separate compilation, row 4 is just more separate. And that separation is between standard modules, which may not be calling into each other that much, as opposed to the user code calling into standard modules (?).

> This is starting to get into what I was going to open as a separate issue, but I think we should potentially do this for this release, as I think it's very low-hanging fruit and very useful. It won't actually help the CHAMPS team since they'll probably still want to stop the compilation after the first step,

An issue would be great. Before going into too much detail here, I find it more likely that CHAMPS can make use of such a flag. Considering the last row using separate compilation can reduce the memory footprint from 3.7G to 3.1G, which is not a huge drop. But yeah... just another guess :)

**Comment by lydia-duncan** [comment]
> > so any improvements to --incremental would be a passing benefit.
> 
> That's a very patient / long-view perspective given that users are suffering from (and complaining about) compile times today and a compiler that supports separate compilation / incremental compilation is still arguably years away. Particularly when the fix literally took me 10 minutes to implement. :D

To be clear, I'm not trying to argue against improvements that are already being worked on or ones that are quick to do.

**Comment by bradcray** [comment]
@lydia-duncan / @e-kayrakli : Before going to the group to ask about the design of this PR, I wanted to make sure:  Are you both in favor of this PR going in as-is without any need for a feature to fall back on the old/current behavior?

**Comment by lydia-duncan** [comment]
Yeah, that seems reasonable to me

**Comment by e-kayrakli** [comment]
I'm ok with this going in as-is.

How many users are dependent on the old behavior? How many users are aware of the old behavior? I think the answer is 0 for both. Nor I see any danger in this implementation that would introduce a bug.

**Comment by bradcray** [comment]
> How many users are dependent on the old behavior? How many users are aware of the old behavior? I think the answer is 0 for both. Nor I see any danger in this implementation that would introduce a bug.

I agree, but since it's a last-minute change, didn't want to be viewed as sneaking it past anyone.

**Comment by bradcray** [comment]
For future reference.  It seems I made a mistake and merged fd48ecd6daa4f2b40ad984cf8a5e5b6f1ead8d98 into this #19407 at some point accidentally, probably when testing to see how the two of them worked together.  I didn't notice this until I'd merged that PR and found that this one was merged as well.  Sorry for the sloppiness and any resulting confusion.

**Pull request merged**
